### ISSUE TITLE
hkdf::KeyType implementation on hmac::Algorithn uses output length rather than block length?

### ISSUE THREAD

**Issue by DoumanAsh** [briansmith/ring#1828]
At the current moment HKDF `KeyType` defines length of key as follows: https://github.com/briansmith/ring/blob/main/src/hmac.rs#L259-L263

```rust
impl hkdf::KeyType for Algorithm {
    fn len(&self) -> usize {
        self.digest_algorithm().output_len()
    }
}
```

But when I look at [hmac::Key::new](https://docs.rs/ring/0.17.5/ring/hmac/struct.Key.html#method.new) it specifies requirements on key length as follows:

```
As specified in RFC 2104, if key_value is shorter than the digest algorithm’s block length (as returned by digest::Algorithm::block_len(), not the digest length returned by digest::Algorithm::output_len()) then it will be padded with zeros. Similarly, if it is longer than the block length then it will be compressed using the digest algorithm.

You should not use keys larger than the digest_alg.block_len because the truncation described above reduces their strength to only digest_alg.output_len * 8 bits. Support for such keys is likely to be removed in a future version of ring.
```

So given example of SHA 512 digest, its block length is 128 bytes while output length is 64 bytes.
Which makes me wonder why it is not block length as one would expect from doc string? Is zero padding non-issue?

I'm raising this issue only because I was trying to find reference on how one should generate HMAC key using HKDF algorithm within ring documentation (cuz I'm not really expert in this stuff)
Obviously I should not use existing `KeyType` as reference and instead opt out to use implement my own struct which will return  block length of my algorithm (I'm not sure what is impact of zero padding)
But it makes me wonder what is the reason for current implementation of `hkdf::KeyType` for `hmac::Algorithm`?

**Comment by briansmith** [comment]
I agree with your thinking in the abstract. However, when I looked at relevant standards (TLS 1.3 and others) I found that almost all standards recommend or require that a HMAC with an N-bit output should have an N-bit key. Thus if we chose the better choice, we'd have something useless to most people because the wouldn't be able to use it for implementing most standard things.

However, you are not the first person to question this choice.  Perhaps we were trying to be too fancy with this API. Perhaps we should remove this kind of convenience feature.

**Comment by DoumanAsh** [comment]
Thanks for clarification
I do not think it is per se needs to be removed, but adding some clarification on the choice of default implementation would be good
In fact just stating it in https://docs.rs/ring/0.17.6/ring/hkdf/trait.KeyType.html would be more than enough and also serve as good guidance to anyone who would want to customize length themself

I wasn't aware of TSL & friends recommendation for example
